source/client_channel_handler.c

```diff
@@ -819,27 +822,53 @@ uint16_t mqtt_create_request(
             aws_raise_error(AWS_ERROR_MQTT_NOT_CONNECTED);
             return 0;
         }
+        /**
+         * Find a free packet ID.
+         * QoS 0 PUBLISH packets don't actually need an ID on the wire,
+         * but we assign them internally anyway just so everything has a unique ID.
+         *
+         * Yes, this is an O(N) search.
+         * We remember the last ID we assigned, so it's O(1) in the common case.
+         * But it's theoretically possible to reach O(N) where N is just above 64000
+         * if the user is letting a ton of un-ack'd messages queue up
```

Comment:
Previously, we'd start searching from 0 each time for a free ID. This made things noticeably slow when the buffer was getting full. This makes it way faster. If anyone can think of a clever solution that isn't O(N) we can implement it in a follow-up

Reply:
As long as we're using a hash table, it's going to be tough to make this clever, other than to maintain a linear array of freed ids and use that once the table has been saturated once.